File: extractor/url_safety.py
"""URL validation and SSRF defenses for the extractor service.

Since this service accepts arbitrary user-supplied URLs and fetches them
server-side, every URL must be validated for scheme and resolved-address
safety before any network request is made. Validation happens against the
*resolved* IP addresses (not just the hostname) to defend against DNS
rebinding, per the planner's SSRF guidance.
"""
import ipaddress
import socket
from urllib.parse import urlparse

ALLOWED_SCHEMES = {"http", "https"}


class UnsafeURLError(ValueError):
    def __init__(self, code: str, message: str):
        self.code = code
        self.message = message
        super().__init__(message)


def _is_blocked_ip(ip_str: str) -> bool:
    try:
        ip = ipaddress.ip_address(ip_str)
    except ValueError:
        return True
    return (
        ip.is_private
        or ip.is_loopback
        or ip.is_link_local
        or ip.is_multicast
        or ip.is_reserved
        or ip.is_unspecified
    )


def validate_public_http_url(raw_url: str) -> str:
    raw_url = (raw_url or "").strip()
    if not raw_url:
        raise UnsafeURLError("INVALID_URL", "url is required")

    parsed = urlparse(raw_url)
    if parsed.scheme not in ALLOWED_SCHEMES:
        raise UnsafeURLError("INVALID_URL", "only http/https URLs are supported")
    if not parsed.hostname:
        raise UnsafeURLError("INVALID_URL", "url is missing a host")
    if parsed.username or parsed.password:
        raise UnsafeURLError("INVALID_URL", "credentials in url are not allowed")

    try:
        addrinfo = socket.getaddrinfo(parsed.hostname, None)
    except socket.gaierror as err:
        raise UnsafeURLError("INVALID_URL", f"could not resolve host: {err}") from err

    resolved_ips = {info[4][0] for info in addrinfo}
    if not resolved_ips:
        raise UnsafeURLError("INVALID_URL", "could not resolve host")

    for ip in resolved_ips:
        if _is_blocked_ip(ip):
            raise UnsafeURLError(
                "INVALID_URL", "url resolves to a disallowed network address"
            )

    return raw_url

File: api/url_safety.go
package main

import (
	"fmt"
	"net"
	"net/url"
	"strings"
)

// ErrUnsafeURL is returned when a candidate download URL is malformed or
// resolves to a network address the service must not reach out to.
type ErrUnsafeURL struct {
	Message string
}

func (e *ErrUnsafeURL) Error() string {
	return e.Message
}

// validateDownloadURL defends the first entry point for user-supplied URLs
// against SSRF: it requires http/https, rejects embedded credentials, and
// rejects any URL whose resolved address falls in a private, loopback,
// link-local, or otherwise reserved range. Validation happens after DNS
// resolution (not just on the hostname string) to defend against DNS
// rebinding, per the planner's SSRF guidance.
func validateDownloadURL(rawURL string) error {
	rawURL = strings.TrimSpace(rawURL)
	if rawURL == "" {
		return &ErrUnsafeURL{Message: "url is required"}
	}

	parsed, err := url.Parse(rawURL)
	if err != nil {
		return &ErrUnsafeURL{Message: "url could not be parsed"}
	}

	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return &ErrUnsafeURL{Message: "only http/https urls are supported"}
	}

	if parsed.User != nil {
		return &ErrUnsafeURL{Message: "credentials in url are not allowed"}
	}

	host := parsed.Hostname()
	if host == "" {
		return &ErrUnsafeURL{Message: "url is missing a host"}
	}

	ips, err := net.LookupIP(host)
	if err != nil {
		return &ErrUnsafeURL{Message: fmt.Sprintf("could not resolve host: %v", err)}
	}
	if len(ips) == 0 {
		return &ErrUnsafeURL{Message: "could not resolve host"}
	}

	for _, ip := range ips {
		if isBlockedIP(ip) {
			return &ErrUnsafeURL{Message: "url resolves to a disallowed network address"}
		}
	}

	return nil
}

func isBlockedIP(ip net.IP) bool {
	if ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() ||
		ip.IsLinkLocalMulticast() || ip.IsMulticast() || ip.IsUnspecified() {
		return true
	}

	// Documentation/reserved ranges and the IPv4 cloud metadata address are
	// covered by IsPrivate/IsLinkLocalUnicast above (169.254.0.0/16), but be
	// explicit for defense in depth.
	if ip4 := ip.To4(); ip4 != nil {
		if ip4[0] == 0 { // 0.0.0.0/8
			return true
		}
	}

	return false
}
